Source/Core/Common/Debug/CodeTrace.cpp

```diff
@@ -0,0 +1,394 @@
+  // individually.
+  return std::make_tuple(count, empty_target, timed_out);
+}
+
+const HitType CodeTrace::TraceLogic(std::vector<std::string>& reg_tracked,
+                                    std::set<u32>& mem_tracked, TraceOutput& current_instr,
+                                    bool first_hit)
+{
+  // Tracks the original value that is in the targeted register or memory through loads, stores,
+  // register moves, and value changes. Also finds when it is used. ps operations are not fully
+  // supported. -ux memory instructions may need special cases.
+  // Should not be called if reg and mem tracked are empty.
+
+  // Using a std::set because it can easily insert the memory range being accessed without
+  // causing duplicates, and quickly erases all members of the memory range without caring if the
+  // element actually exists.
+
+  bool mem_hit = false;
+  if (current_instr.memory_target && !mem_tracked.empty())
+    mem_hit = CompareMemoryTargetToTracked(current_instr.instruction,
+                                           current_instr.memory_target.value(), mem_tracked);
+
+  // Optimization for tracking a memory target when no registers are being tracked.
+  if (reg_tracked.empty() && !mem_hit)
+    return HitType::SKIP;
+
+  // Break instruction down into parts to be analyzed.
+  InstructionAttributes instr = GetInstructionAttributes(current_instr);
+
+  // Not an instruction we care about (branches).
+  if (instr.reg0.empty())
+    return HitType::SKIP;
+
+  // The itR will be used later for erasing.
+  auto itR = std::find(reg_tracked.begin(), reg_tracked.end(), instr.reg0);
+  const bool match_reg123 =
+      (!instr.reg1.empty() &&
+       std::find(reg_tracked.begin(), reg_tracked.end(), instr.reg1) != reg_tracked.end()) ||
+      (!instr.reg2.empty() &&
+       std::find(reg_tracked.begin(), reg_tracked.end(), instr.reg2) != reg_tracked.end()) ||
+      (!instr.reg3.empty() &&
+       std::find(reg_tracked.begin(), reg_tracked.end(), instr.reg3) != reg_tracked.end());
+  const bool match_reg0 = (itR != reg_tracked.end());
+
+  if (!match_reg0 && !match_reg123 && !mem_hit)
+    return HitType::SKIP;
+
+  // Exclusions from updating tracking logic. mt operations are too complex and specialized.
+  // Combiner used later.
+  const std::vector<std::string> exclude{"dc", "ic", "mt"};
+  const std::vector<std::string> compare{"c", "fc"};
+
+  // rlwimi, at least, can preserve parts of the target register. Not sure if rldimi can too or if
+  // there are any others like this.
+  const std::vector<std::string> combiner{"rlwimi"};
+
+  // Link register for r0 tracking
+  if (StringBeginsWith(instr.instruction, "mflr") && match_reg0)
+  {
+    reg_tracked.erase(itR);
+    return HitType::OVERWRITE;
+  }
+  else if (StringBeginsWith(instr.instruction, "mtlr") && match_reg0)
+  {
+    // LR is not something tracked
+    return HitType::MOVED;
+  }
+
+  if (CompareInstruction(instr.instruction, exclude))
```

Comment:
If `CompareInstruction` was either a lambda with an `auto` parameter inside, or a template function with the size then the `exclude`/`compare`/`combiner` could be `std::array<std::string_view, N>`

Ex:

```
const auto CompareInstruction = [](std::string_view instruction, const auto& type_compare)
{
  return std::any_of(type_compare.begin(), type_compare.end(), [&instruction](std::string_view s){ return StringBeginsWith(instruction, s); });
};


// Exclusions from updating tracking logic. mt operations are too complex and specialized.
// Combiner used later.
const std::array<std::string_view, 3> exclude{"dc", "ic", "mt"};
const std::array<std::string_view, 2> compare{"c", "fc"};

// rlwimi, at least, can preserve parts of the target register. Not sure if rldimi can too or if
// there are any others like this.
const std::array<std::string_view, 1> combiner{"rlwimi"};

....
if (CompareInstruction(instr.instruction, exclude))
{
  ...
}
else if (CompareInstruction(instr.instruction, compare))
{
  ...
}
...
```

This has the advantage of avoiding the `std::vector` which generates dynamic memory at runtime.